[5.1] Enable Router Switching

Keoghan opened the pull request:
Hi, 

I'm pretty early in my Laravel career - so I apologise if my contribution is not quite of the correct format. I'll be happy to amend as/if necessary.

I've proposed some modifications to Http Kernel, applying middleware after bootstrapping, and retrieving the router from the container later, therefore allowing a service provider to override the router.

This also enables a slight tidy up where the router no longer needs to be aware of the container settings for applying middleware - we just don't set up the middleware in the Kernel.

Keoghan committed:
Enable Router Switching

Modifications to Http Kernel, applying middleware after bootstrapping,
and retrieving the router from the container later, therefore allowing
a service provider to override the router.

Also enables a slight tidy up where the router no longer needs to be
aware of the container and whether to apply middleware - we just don't
set up the middleware in the Kernel.

GrahamCampbell commented on src/Illuminate/Routing/Router.php:
```diff
@@ -848,6 +845,23 @@ public function middleware($name, $class)
+     * @param array $middlewares
+     * @return $this;
+     */
+
```
no blank line please

GrahamCampbell commented on src/Illuminate/Routing/Router.php:
```diff
@@ -848,6 +845,23 @@ public function middleware($name, $class)
     }
 
     /**
+     * Bulk register short-hand names for a set of middleware. Middlewares
```
Please start the next sentence two lines down.

Keoghan committed:
Tidy up.

Remove unused Illuminate\Routing\Router declaration.

Clean up code formatting as requested.

lukasgeiter commented on src/Illuminate/Routing/Router.php:
```diff
@@ -848,6 +845,23 @@ public function middleware($name, $class)
+     *
+     * Middlewares are passed through with the format [ 'name' => 'class', ... ]
+     *
+     * @param array $middlewares
```
Give these a bit more space ;) There should be two spaces between them `@param  array  $middlewares`

lukasgeiter commented on src/Illuminate/Routing/Router.php:
```diff
@@ -848,6 +845,23 @@ public function middleware($name, $class)
+     * Middlewares are passed through with the format [ 'name' => 'class', ... ]
+     *
+     * @param array $middlewares
+     * @return $this;
```
No need for a semicolon here

Keoghan committed:
Tidy up PR

Formatting changes to DocBlock for PR.

GrahamCampbell commented:
What's your use case for this?

Keoghan commented:
Hi Graham, 

I was trying to use route-model binding in a clean manner for using slugs on a multiple number of models. The cleanest way I could find was to override the `model()` method on the router - where I could make it search by either slug, or id. (Details on konsulting/eloquent-sluggable). 

At the moment I have manually hacked the Http Kernel for my project - but having read some conversations on Laravel.io (e.g. http://laravel.io/forum/02-12-2015-l5-extending-router) and a few around the web where people were trying to use a customised router in L5 but couldn't (and seemingly could in L4), I thought I'd share :)

GrahamCampbell commented:
You just need to bind the router object before your instantiate the http kernel.

Keoghan commented:
Thanks Graham, I see that I could do that within /bootstrap/app.php.

Being new to the scene, I don't know if it is a common thing to edit the bootstrap/app.php file - I'd felt that if you could use a service provider, it would be more in keeping with the approaches I'd seen so far.

Appreciate you taking a look anyway.

GrahamCampbell commented:
It's not common, but overriding the router isn't common, so probably fine. :)
